tokio/src/util/wake_list.rs

```diff
@@ -0,0 +1,44 @@
+use core::mem::{self, MaybeUninit};
+use std::task::Waker;
+
+const NUM_WAKERS: usize = 32;
+
+pub(crate) struct WakeList {
+    inner: [MaybeUninit<Waker>; NUM_WAKERS],
+    curr: usize,
+}
+
+impl WakeList {
+    pub(crate) fn new() -> Self {
+        Self {
+            inner: unsafe { MaybeUninit::uninit().assume_init() },
+            curr: 0,
+        }
+    }
+
+    pub(crate) fn can_push(&self) -> bool {
+        self.curr < NUM_WAKERS - 1
+    }
+
+    pub(crate) fn push(&mut self, val: Waker) {
+        debug_assert!(self.can_push());
+
+        self.inner[self.curr] = MaybeUninit::new(val);
+        self.curr += 1;
+    }
+
+    pub(crate) fn wake_all(&mut self) {
+        for waker in &mut self.inner[..self.curr] {
+            unsafe { mem::replace(waker, MaybeUninit::uninit()).assume_init() }.wake()
+        }
+        self.curr = 0;
```

Comment:
Here if `wake` were to panic `Drop` would be also run on uninitialized wakers

Reply:
Yes, `wake` can panic. It is user-supplied code.

Reply:
I see this code has changed in later commits, but why would `wake` panicking cause `Drop` to be called on uninit wakers? The wakers are wrapped in `MaybeUninit`, which doesn't call `Drop` on its inner value.  Just trying to understand the original comment here.

Reply:
Wakers must be properly Dropped even if panic occurred on wake attempt. In order to do it, a custom destructor is implemented on WakeList, which takes wakers out of MaybeUninit and make them correctly drop. But this implementation could lead to `self.curr` not being updated after panic, and Drop implementation will try to take already taken weaker, which already became uninitialized.

Reply:
> The wakers are wrapped in `MaybeUninit`, which doesn't call `Drop` on its inner value.

Right

> I see this code has changed in later commits, but why would `wake` panicking cause `Drop` to be called on uninit wakers? 

A panic before `self.curr = 0` would cause the `Drop` implementation of `WakeList` to read and drop items of `self.inner` which could have been already been dropped or even de-initialized. This type of issue is explained better here https://doc.rust-lang.org/nomicon/exception-safety.html

Reply:
Ah, I had missed that `WakeList` had a custom `Drop` impl that was dropping the wakers.  All makes sense now, thanks.  I had for a second thought I was misunderstanding something (subtle) about `MaybeUninit`.